### ISSUE TITLE
Investigate possible BUG Invoke-WebRequest content type charset

### ISSUE THREAD

**Issue by CarloToso** [PowerShell/PowerShell#19156]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Invoke-WebRequest POST with -Body (string) and -Headers containing a content type with charset (ex. plain/text; charset="windows-1251") and no -ContentType.

```ps1
Invoke-WebRequest -URI "some uri" -Method POST -Headers @{Content-Type="plain/text; charset=windows-1251"} -Body "string encoded with windows-1251 charset"
```

The charset in the headers are ignored be ignored and default to UTF8.
This doesn't happen if we use -ContentType

```ps1
Invoke-WebRequest -URI "some uri" -Method POST -ContentType "plain/text; charset=windows-1251" -Body "string encoded with windows-1251 charset"
```

Both should have the same behaviour

I found this possible bug while reading the code; I haven't encountered it in the real world so I don't really know how we could test it (and if testing it is even possible)

### Expected behavior

```console
SetRequestContent should try to get the charset from the ContentType in Headers saved in the WebSession
WebSession.ContentHeaders[HttpKnownHeaderNames.ContentType]
```

### Actual behavior

```console
SetRequestContent only checks for -ContentType charset
```

### Error details

```console
While writing some code in `WebRequestPSCmdlet.Common.cs` think I found a BUG (line 1491) it should check for `WebSession.ContentHeaders[HttpKnownHeaderNames.ContentType]` instead of `ContentType` (this should be ok because if both are present `ContentType` overwrites `WebSession.ContentHeaders[HttpKnownHeaderNames.ContentType]`) unfortunately I don't know how to test it (I think it would require sending a wierdily encoded string as Body), if confirmed I'll open another PR to fix it.
```

### Environment data

```powershell
Latest master
https://github.com/PowerShell/PowerShell/blob/60b4e9704fca546f76e1128e30de4077c51a3ca2/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
```

**Comment by iSazonov** [comment]
@CarloToso Please make more clear repro steps and description.

**Comment by mklement0** [comment]
If I understand correctly, the problem is that a `charset` value is only honored if provided via `-ContentType`, not also as part of `-Headers`.

Source-code link:

https://github.com/PowerShell/PowerShell/blob/27313be95bab73f3066ab7382240d72e2572dddb/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1491-L1503

**Comment by CarloToso** [comment]
@mklement0 Exactly, but I don't know how to test it

**Comment by mklement0** [comment]
@CarloToso, you're right - I misread the source code; I've edited my previous comment to remove the mistaken interpretation.

**Comment by rhubarb-geek-nz** [comment]
The [documentation ](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest?view=powershell-7.3) describes the behaviour so it is not a bug, it is a feature.

```
f the value for ContentType contains the encoding format (as charset), the cmdlet uses that format to encode the body of the web request. If the ContentType doesn't specify an encoding format, the default encoding format is used instead.
```

**Comment by mklement0** [comment]
@rhubarb-geek-nz, without weighing in on what the right resolution is in this particular case, let me make a general point:

* It is helpful to point out what the _documented behavior_ is.

* It is _not_ helpful to proclaim that the documented behavior is _ipso facto_ the correct one.

Whether the current behavior ends up being deemed (a) a _bug_ or (b) _requiring improvement_ or (c) _neither_ is a separate matter, which needs to be debated on the merits of the arguments made.

**Comment by rhubarb-geek-nz** [comment]
By feature I meant that the documentation is consistent with the implementation.

**Comment by mklement0** [comment]
@rhubarb-geek-nz, the premise of this issue is that _only_ `-ContentType` is honored, which is the (only) part covered by the documentation. As such, your comment doesn't address the specific issue being raised: why isn't an - arguably equivalent - `charset` definition _also_ honored as part of the `-Headers` argument?
It's fine to say, "here's what the docs say"; it's not helpful to say, "it's not a bug, it's a feature", without addressing the arguments made that describe the current behavior as deficient.

**Comment by rhubarb-geek-nz** [comment]
The current documentation is suggesting that in order to use a character set other than the default one, then one should use the ContentType argument rather than rely on the headers being interpreted. This is a consistent approach and avoids the ambiguity of multiple headers, or a content type where the character set is implied by the content type but not explicitly encoded in the literal.

**Comment by mklement0** [comment]
>  rather than rely on the headers being interpreted. 

That part is your conjecture.
And even if this _was_ what the docs meant to say, at the _very least_ the docs should be amended to _state that explicitly_, especially since it is reasonable to conceive of `-ContentType` as a _shortcut_ to a `Content-Type` attribute in a hashtable passed to `-Headers`.

Of course, a different resolution could be to implement - and document - _precedence rules_.

Again: I'm not weighing in on what the correct resolution is; I'm trying to point what is required for a constructive, nuanced debate (which shouldn't require several comments just to draw out what your original comment was meant to express).

**Comment by rhubarb-geek-nz** [comment]
> The charset in the headers should be ignored and default to UTF8
> 
> I found this possible bug while reading the code; I haven't encountered it in the real world so I don't really know how we could test it (and if testing it is even possible)

I wrote a small web server that returns the raw request body as hex. This allows us examine what the behaviour is. My primary concern was not about whether it should also interpret the headers but what actual encoding rules currently are.

Test client talks to http://localhost:8080 with POST and various sample data. A single German Eszett character was used because the encoding is different between ISO-8859-1 and UTF-8.

```
#!/usr/bin/env pwsh

"Default encoding" | Write-Host

[System.Text.Encoding]::Default.BodyName | Write-Host

$ESZETT = "`u{00DF}"

"Raw character" | Write-Host

$ESZETT | Write-Host

"Default encoding" | Write-Host

Invoke-WebRequest -Uri http://localhost:8080 -Method POST -Body $ESZETT | Write-Host

"As iso-8859-1" | Write-Host

Invoke-WebRequest -Uri http://localhost:8080 -Method POST -Body $ESZETT -ContentType "plain/text; charset=iso-8859-1" | Write-Host

"As UTF-8" | Write-Host

Invoke-WebRequest -Uri http://localhost:8080 -Method POST -Body $ESZETT -ContentType "plain/text; charset=utf-8" | Write-Host

"As binary encoding with UTF8" | Write-Host

$bytes = [System.Text.Encoding]::UTF8.GetBytes($ESZETT)

Invoke-WebRequest -Uri http://localhost:8080 -Method POST -Body $bytes | Write-Host

"JSON encoding" | Write-Host

$dict = @{S=$ESZETT}

$json = ConvertTo-Json -InputObject $dict

$json | Write-Host

"As JSON string" | Write-Host

Invoke-WebRequest -Uri http://localhost:8080 -Method POST -Body $json -ContentType "application/json" | Write-Host

"As JSON bytes" | Write-Host

$bytes = [System.Text.Encoding]::UTF8.GetBytes($json)

Invoke-WebRequest -Uri http://localhost:8080 -Method POST -Body $bytes -ContentType "application/json" | Write-Host
```

Results with commentary

```
Default encoding
utf-8
```

This is a good start, because now we know what the default character set is. Except, no, we don't. On .NET Core UTF-8 is always returned as the default character set, and as we will see is not the default used by Invoke-WebRequest.

```
Raw character
ß
```

This is a good result, the encoding made it all the way to the console output.

```
Default encoding
DF 
```

So we now know that the default encoding is definitely not UTF-8. But we don't know what it is or where is comes from.

```
As iso-8859-1
DF
```

The default encoding may be ISO-8859-1 but we don't know for sure.

```
As UTF-8
C3 9F
```

We have confirmed now that ContentType is respected and used in the translation

```
As binary encoding with UTF8
C3 9F
```

This shows if required we can do the translation ourselves.

```
JSON encoding
{
  "S": "ß"
}
```

Confirmation of JSON encoding with our test character

```
As JSON string
7B 0A 20 20 22 53 22 3A 20 22 DF 22 0A 7D
```

Unfortunately this is a bad result. The application/json type is defined by IANA as a [binary](https://www.iana.org/assignments/media-types/application/json) format. JSON.org says it is always encoded using UNICODE. So you should not need to add a charset to application/json content type, and a web server may not be expecting a charset with application/json, because it is defined as a binary format.

```
As JSON bytes
7B 0A 20 20 22 53 22 3A 20 22 C3 9F 22 0A 7D
```

This is good, we can specify the content type using the argument and ensure the encoding is correct ourselves.

In summary

- The tests don't confirm what the default character encoding is, but we know it is not UTF-8
- If you are exchanging JSON or XML then you may have to do the encoding yourself rather than rely on Invoke-WebRequest

**Comment by CarloToso** [comment]
@rhubarb-geek-nz what version of powershell are you using for your tests?

**Comment by rhubarb-geek-nz** [comment]
> @rhubarb-geek-nz what version of powershell are you using for your tests?

7.3.3

**Comment by CarloToso** [comment]
The default encoding for Invoke-WebRequest in 7.3 is iso-8859-1, changed to utf-8 in 7.4 preview1

**Comment by rhubarb-geek-nz** [comment]
> The default encoding for Invoke-WebRequest in 7.3 is iso-8859-1, changed to utf-8 in 7.4 preview1

That will certainly help common JSON and XML cases.

**Comment by rhubarb-geek-nz** [comment]
Using powershell-7.4.0-preview.1-osx-x64 

```
Default encoding
utf-8
Raw character
ß
Default encoding
C3 9F                                                                                                                   
As iso-8859-1
DF
As UTF-8
C3 9F
As binary encoding with UTF8
C3 9F
JSON encoding
{
  "S": "ß"
}
As JSON string
7B 0A 20 20 22 53 22 3A 20 22 C3 9F 22 0A 7D
As JSON bytes
7B 0A 20 20 22 53 22 3A 20 22 C3 9F 22 0A 7D
```

So this is consistent with using UTF-8 as default

**Comment by CarloToso** [comment]
There is also a special workaround (that can maybe be removed), that sets the encoding to utf8 when the body is json (I will post the link to the code here later)

**Comment by rhubarb-geek-nz** [comment]
Other cmdlets use an "-Encoding" parameter, eg Get-Content. It would be nice if this was used for the default encoding character set when the character set is not present in the ContentType and the body is some form of string rather than binary. This would be better than treating certain MIME types as somehow special.

**Comment by CarloToso** [comment]
Maybe you could open a new issue to track that.

**Comment by mklement0** [comment]
@CarloToso, 

Based on my summary of what I think the issue is, and your confirmation of it as correct,  I'm confused by the following sentence in the initial post:

> The charset in the headers should be ignored and default to UTF8

To reconfirm: It should _not_ be ignored, and should instead be honored, just like it already is for `-ContentType`, correct?

---

Proceeding on the assumption that my summary is correct:

A simple, setup-free way to verify that `-ContentType` _is_ honored, but an equivalent `Content-Type` header field is _not_ is via https://httpbin.org/post, which simply echoes the body back to you:

* Given that UTF-8 is JSON's default, and that PowerShell now uses UTF-8 encoding by the default for a `[string]`-based `-Body` argument with content type `application/json`, the following works as expected:

```powershell
# Output: [pscustomobject] @{ foo = 'sß' }
(Invoke-RestMethod https://httpbin.org/post -Body '{ "foo": "sß" }' -Method POST -ContentType application/json).json
```

* If you send the _wrong_ encoding, which PowerShell applies due to the `charset` attribute in the `-ContentType` argument, the `.json` property value is empty:

```powershell
# NO output, because the effectively ISO-8859-1-encoded body cannot be decoded *as UTF8* by https://httpbin.org/post
(Invoke-RestMethod https://httpbin.org/post -Body '{ "foo": "sß" }' -Method POST -ContentType 'application/json; charset=iso-8859-1').json
```

* Trying to make the same intentional mistake with an equivalent `Content-Type` header field does NOT provoke the symptom, which implies that the `charset` attribute was _ignored_  by PowerShell, and that UTF-8 was sent (the attribute is seemingly also ignored by https://httpbin.org/post):

```powershell
# !! Output: [pscustomobject] @{ foo = 'sß' } - despite the attempt to send broken encoding
(Invoke-RestMethod https://httpbin.org/post -Body '{ "foo": "sß" }' -Method POST -Headers @{ 
  'Content-Type' = 'application/json; charset=iso-8859-1' 
}).json
```

**Comment by CarloToso** [comment]
@mklement0 Yes exactly, I'll edit the initial post to make it clearer
